clients/include/testing_gemm_ex.hpp

```diff
@@ -241,6 +241,61 @@ void testing_gemm_ex_bad_arg(const Arguments& arg)
                           rocblas_status_invalid_handle);
 }
 
+template <typename Ti, typename To, typename Tc>
+void reference_gemm(rocblas_operation transA,
+                    rocblas_operation transB,
+                    rocblas_int       m,
+                    rocblas_int       n,
+                    rocblas_int       k,
+                    Tc                alpha,
+                    Ti*               A,
+                    rocblas_int       lda,
+                    Ti*               B,
+                    rocblas_int       ldb,
+                    Tc                beta,
+                    To*               C,
+                    rocblas_int       ldc)
+{
+    cblas_gemm<Ti, To, Tc>(transA, transB, m, n, k, alpha, A, lda, B, ldb, beta, C, ldc);
+}
+
+namespace
+{
+    bool is_replacement_kernel(rocblas_int m, rocblas_int n)
+    {
+        if((m == 512 && n == 512) || (m == 1024 && n == 1024) || (m == 2048 && n == 2048)
+           || (m == 4096 && n == 4096) || (m == 960 && n == 1024) || (m == 3840 && n == 4096))
+            return true;
+        return false;
+    }
+}
+
+template <>
+void reference_gemm(rocblas_operation transA,
+                    rocblas_operation transB,
+                    rocblas_int       m,
+                    rocblas_int       n,
+                    rocblas_int       k,
+                    float             alpha,
+                    rocblas_bfloat16* A,
+                    rocblas_int       lda,
+                    rocblas_bfloat16* B,
+                    rocblas_int       ldb,
+                    float             beta,
+                    rocblas_bfloat16* C,
+                    rocblas_int       ldc)
+{
+    const size_t       size_C = size_t(ldc) * size_t(n);
+    host_vector<float> C_float(size_C);
+    for(int i = 0; i < size_C; ++i)
+        C_float[i] = C[i];
+    cblas_gemm<rocblas_bfloat16, float, float>(
+        transA, transB, m, n, k, alpha, A, lda, B, ldb, beta, C_float, ldc);
+    bool round = !is_replacement_kernel(m, n);
+    for(int i = 0; i < size_C; ++i)
+        C[i] = rocblas_bfloat16(C_float[i], round);
+}
+
```

Comment:
This `reference_gemm` needs to support all combinations of `rocblas_gemm_ex`, `rocblas_batched_ex`, and `rocblas_strided_batched_ex`.

Is there any reason `cblas_gemm` is not adequate, even if we convert to `float` first?

Reply:
cblas_gemm<bfloat16, bfloat16, float> rounds the answer. cblas_gemm<bfloat16, float, float> works, but calling it is a special case since To doesn't match the test case To. The alternative is to write a specialized version of testing_gemm_ex which would result in more duplicated code. The reference_gemm as a wrapper for cblas_gemm has less duplicated code, and should support all combinations of rocblas_gemm_ex through templates.